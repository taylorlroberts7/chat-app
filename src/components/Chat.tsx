import * as React from 'react';
import { database } from '../firebase';
import { Messages } from './index';

export default class Chat extends React.Component<null, IChatState> {

    state: IChatState = {
        error: null,
        newMessage: null,
        messages: []
    }

    _inputRef = React.createRef<HTMLTextAreaElement>();

    componentDidMount() {

        const messagesRef = database.ref('messages')
            .orderByKey()
            .limitToLast(100);

        messagesRef.on('value', snapshot => {

            let messagesObj = snapshot.val();
            let messages: any[] = [];

            if (messagesObj) {

                Object.keys(messagesObj)
                    .forEach(key => messages.push({
                        ...messagesObj[key]
                    }));

            }

            this.setState({ messages });

        });

    }

    render() {

        const { error, messages } = this.state;

        return (
            <div>
                <Messages messages={messages} />
                <form
                    className="message-form"
                    onSubmit={this._handleSendMessage}
                >
                    <textarea
                        onChange={e => this.setState({ newMessage: e.target.value })}
                        onKeyDown={this._handleEnterPress}
                        ref={this._inputRef}
                    />
                    <button>Send</button>
                </form>
                {error}
            </div>
        )
    }

    _handleEnterPress = (event: React.KeyboardEvent<HTMLTextAreaElement>) => {

        // This allows users to send message when enter key is pressed and shift key is not held down
        if (event.keyCode === 13 && !event.shiftKey) {

            event.preventDefault();
            return this._handleSendMessage();

        }

        return;

    }

    _handleSendMessage = (event?: React.FormEvent) => {

        event && event.preventDefault();

        const { newMessage } = this.state;

        if (!newMessage) {

            return;

        }

        try {

            const username = window.sessionStorage.getItem('username');

            const value = {
                text: newMessage,
                sentAt: Date.now(),
                user: username
            };

            // Add message to DB
            database.ref('messages').push(value);

            // Clear message input
            this._inputRef.current!.value = '';

        } catch (e) {

            // TODO: Update error message
            const error = 'Woopsies';
            this.setState({ error });

        }

    }

}

interface IChatState {
    error: string | null;
    newMessage: string | null;
    messages: IMessage[];
}

export interface IMessage {
    text: string;
    sentAt: number;
    user: string;
}
