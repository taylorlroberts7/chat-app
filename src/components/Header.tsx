import * as React from 'react';
import { History } from 'history';

const Header: React.FC<IHeaderProps> = ({ history }) => {

    return (
        <div className="header">
            <button onClick={() => handleClearUsername(history)}>
                Change Username
            </button>
        </div>
    );

};

const handleClearUsername = (history: History) => {

    window.sessionStorage.removeItem('username');
    history.push('/');

};

export default Header;

interface IHeaderProps {
    history: History;
}
