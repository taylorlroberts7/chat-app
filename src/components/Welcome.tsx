import * as React from 'react';
import { RouteComponentProps } from 'react-router';

const { useState } = React;

const Welcome: React.FC<RouteComponentProps> = ({ history }) => {

    const [error, setError] = useState<string | null>(null);
    const [username, setUsername] = useState<string | null>(null);

    return (
        <div className="welcome">
            <div>
                <h1>Welcome</h1>
                <span>Enter username to chat it up &darr;</span>
            </div>
            <form
                className="username-form"
                onSubmit={e => handleUsernameSubmit(e, username, history, setError)}
            >
                <input
                    aria-label="Username"
                    onChange={e => handleUsernameChange(e, setUsername, setError)}
                    placeholder="username"
                    type="text"
                />
                <span>{error}</span>
                <button>Enter Chat</button>
            </form>
        </div>
    );

};

const handleUsernameChange = (
    event: React.ChangeEvent<HTMLInputElement>,
    setUsername: React.Dispatch<React.SetStateAction<string | null>>,
    setError: React.Dispatch<React.SetStateAction<string | null>>
) => {

    setUsername(event.target.value);
    setError(null);

};

const handleUsernameSubmit = (
    event: React.FormEvent,
    username: string | null,
    history: RouteComponentProps['history'],
    setError: React.Dispatch<React.SetStateAction<string | null>>
) => {

    event.preventDefault();

    if (!username) {

        const error = 'Hold up. Please enter a username to talk to the cool cats.';
        return setError(error);

    }

    sessionStorage.setItem('username', username!);
    history.push('/chat');

};

export default Welcome;
