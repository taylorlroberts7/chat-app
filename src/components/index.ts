export { default as Chat } from './Chat';
export { default as Message } from './Message';
export { default as Messages } from './Messages';
export { default as Welcome } from './Welcome';
