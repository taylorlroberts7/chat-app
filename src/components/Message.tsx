import * as React from 'react';
import { IMessage } from './Chat';

const Message: React.FC<IMessageProps> = ({ message }) => {

    const username = window.sessionStorage.getItem('username');
    const isFromCurrentUser = username === message.user;
    const classNames = `message ${isFromCurrentUser ? 'right' : 'left'}`;

    const date = new Date(message.sentAt);
    const dateString = date.toLocaleDateString(undefined, { month: 'long', day: 'numeric', year: 'numeric' });
    const timeString = date.toLocaleTimeString(undefined, { hour: '2-digit', minute: '2-digit' });

    return (
        <div className={classNames}>
            <div className="content">
                <span>{message.text}</span>
            </div>
            <span className="user-info">
                {message.user} @ {dateString} {timeString}
            </span>
        </div>
    );
}

export default Message;

interface IMessageProps {
    message: IMessage;
}
