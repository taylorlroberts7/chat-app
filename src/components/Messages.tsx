import * as React from 'react';
import { Message } from './index';
import { IMessage } from './Chat';

const Messages: React.FC<IMessagesProps> = ({ loading, messages }) => {

    if (loading) {

        return (
            <div className="messages loading">
                Loading messages
            </div>
        );

    }

    return (
        <div className="messages">
            {!messages.length && (
                <span>
                    No messages yet. How about you get this party started?
                        <span aria-label="party popper" role="img"> 🎉</span>
                </span>
            )}
            {messages.map((m: IMessage, i: number) => (
                <Message message={m} key={`${m.user}-${i}`} />
            ))}
        </div>
    );

};

export default Messages;

interface IMessagesProps {
    loading: boolean;
    messages: IMessage[];
}
