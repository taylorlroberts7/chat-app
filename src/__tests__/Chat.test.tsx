import * as React from 'react';
import { render } from '@testing-library/react';
import { Chat } from '../components';
import { getMockRouterProps } from './mocks';

const routerProps = getMockRouterProps({});
const setup = () => render(<Chat {...routerProps} />);

describe('Chat', () => {

    it('Renders without crashing', () => {

        const { container } = setup();

        expect(container.firstChild).toMatchSnapshot();

    });

});
