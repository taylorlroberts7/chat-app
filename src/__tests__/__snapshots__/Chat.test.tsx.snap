// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Renders without crashing 1`] = `
<div
  class="main"
>
  <div
    class="header"
  >
    <button>
      Change Username
    </button>
  </div>
  <div>
    <div
      class="messages loading"
    >
      Loading messages
    </div>
    <form
      class="message-form"
    >
      <textarea />
      <button>
        Send
      </button>
    </form>
  </div>
</div>
`;
